rgblight 255 hue

XScorpion2 opened the pull request:
Converted rgblight to use 255 hue identical to rgb matrix
Converted some rgblight code to use math8

## Description


## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [x] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

Lenbok commented on quantum/rgblight.c:
```diff
@@ -137,33 +135,9 @@ void setrgb(uint8_t r, uint8_t g, uint8_t b, LED_TYPE *led1) {
 void rgblight_check_config(void) {
-  /* Add some out of bound checks for RGB light config */
-
-  if (rgblight_config.mode < RGBLIGHT_MODE_STATIC_LIGHT) {
```
Why doesn't this function check the mode value any more? Surely you need to cap against the maximum permissible mode?

drashna replied:
I was wondering that myself.  However, the answer is actually simple, it's calling this when the values change: 
https://github.com/qmk/qmk_firmware/blob/c98247e3dd2958bd2d8969dc75170e7e2757b895/lib/lib8tion/math8.h#L23-L28

So, if that is the case, this actually handles that, so a manual check isn't needed, in theory.

Lenbok replied:
Hmmm, sure for when values are being incrementally adjusted (BTW, why do apparently unused `rgblight_increase` and `rgblight_decrease` even exist when there are the `rgblight_step*` functions?). But what about if the mode value stored in `rgblight_config` as read out of eeprom or potentially as transferred between halves of a split keyboard is out of range compared to what the firmware thinks the number of modes are? (e.g. the `rgblight_config` struct change here allocates an extra bit to the mode value, so the value read out will be different to what was originally written). 

Also potentially relevant is #5336

XScorpion2 replied:
That was a mistake, mode checks should remain as its range is less than uint8_t range. Will fix

XScorpion2 replied:
Fixed, if #5336 lands before this pr, will update this pr at that time.

drashna commented on lib/lib8tion/lib8tion.h:
```diff
@@ -167,8 +167,8 @@ Lib8tion is pronounced like 'libation': lie-BAY-shun
 
-#define LIB8STATIC __attribute__ ((unused)) static inline
-#define LIB8STATIC_ALWAYS_INLINE __attribute__ ((always_inline)) static inline
+#define LIB8STATIC
```
Wha was this changed?

XScorpion2 replied:
Firmware size decrease. The unused attribute one at least was causing functions to be included in the firmware that were unused.

drashna replied:
okay.

XScorpion2 replied:
backed out this change, will do this in a separate pr.

drashna commented:
I hate to ask, but could you include the changes to this file:
https://github.com/qmk/qmk_firmware/blob/master/quantum/rgblight_list.h

Also, it may make thing more complicated, but it may be a good idea to incorporate this: #5300

XScorpion2 commented:
I also need to take another stab at improving the gradient code as the way it is calculating the hue for each key has some serious rounding errors due which are a bit worse under 255 vs 360 hue. And since hue is now 255 standard across both. It would make sense to rip out the hue to rgb functions into a common code file.

drashna commented:
Well, improving the animations is something that has been discussed for a long time, and that nobody has wanted to do.  So, awesome!

However, just a heads up, there are some a bunch of compilation errors here.

XScorpion2 commented:
I probably won't get to it tonight before my flight, will pick this back up after vacation. Mind flagging as in progress?

drashna commented:
I figured as much. I just wanted to make sure it was mentioned. 

And enjoy!

drashna commented:
![](https://github.trello.services/images/mini-trello-icon.png) [RGB Light feature code changes](https://trello.com/c/aHZBlShB/3-rgb-light-feature-code-changes)

XScorpion2 commented:
@drashna I don't have permission to see that card.

drashna commented:
Sorry, doing organizational stuff for Ergodox EZ/ZSA, and didn't realize that it posted comments by default. 

/me waves hand
Nothing to see here.

XScorpion2 commented:
Merged in #5300 and updated with 255 based hue values. 
Fixed compile errors.
Fixed bad rounding errors in gradient effect.
Switched rgblight to reuse common rgb to hsv code.

XScorpion2 commented:
@drashna I approve of your approval!

drashna commented:
One thing though.  Could a check for the mode be added to the `rgblight_check_config` function be added?  
This way, if the structure change causes issues with the mode, it should be fixed.

XScorpion2 commented:
@drashna I already put the mode check back into check config after @Lenbok pointed out the error

drashna commented:
Ah, yeah, you're right. 

But I ee that the HSV check was removed.  Was this validated elsewhere, now?

XScorpion2 commented:
@drashna HSV checks were moved to the functions that modify values themselves (qadd8 and qsub8), also since they are now all uin8_t, they can't actually go out of bounds, they just wrap. So no need to check for a case that can never happen.

drashna commented:
Okay, sounds good then.

XScorpion2 commented:
Rebased & Compacted commits onto latest master

drashna commented:
Merge conflicts now, sorry.

XScorpion2 commented:
Not sure if I 100% fixed the merge conflict correctly, needs testing

drashna commented:
It looks like it merged mostly correctly.

However, it looks like there is an issue here.  Specifically, if you use my RGB coexistence PR (#5404), and this, and you enable both features, it generates an error. 

```
 |
 | c:/msys64/home/drashna/qmk_utils/avr8-gnu-toolchain/bin/../lib/gcc/avr/8.3.0/../../../../avr/bin/ld.exe: .build/obj_planck_light_drashna/quantum/rgb_matrix.o (symbol from plugin): in function `b_m16_interleave':
 | (.text+0x0): multiple definition of `b_m16_interleave'; .build/obj_planck_light_drashna/quantum/rgblight.o (symbol from plugin):(.text+0x0): first defined here
 | collect2.exe: error: ld returned 1 exit status
```

Currently, the ONLY way to enable both RGB features is with my PR, since it will error out, otherwise ... it errors out because of the ws2812 rgb matrix PR.

XScorpion2 commented:
@drashna go approve and merge https://github.com/qmk/qmk_firmware/pull/5598 to fix that error. Both pelrun and myself have already approved it.

drashna commented:
Yup, I can confirm that fixes the issue I saw. :)

XScorpion2 commented:
Rebased on latest master, any update on merging this?

drashna commented:
Just looking for some good approvals on the PR, to be honest
